### PR TITLE
chore(vrl, schemas): add new `value` crate, including `Kind` type

### PR DESCRIPTION
**Pull request by JeanMertz**
This commit adds a new `value` crate, which — to start — contains the `Kind` type.

This type is inspired by VRL's `TypeDef` kind, but:

- Is completely separate from VRL, or it's `Value` type.
- Has an improved/safer/easier-to-use API.
- Does not include the concept of "fallibility".
- Is properly unit-tested.

The goal of this crate is to share major components of VRL's progressive type checking between VRL and Vector, as Vector also learns about even field types, as part of the ongoing "schema" work.

This PR does not yet integrate the crate into either VRL or Vector. That work will happen in follow-up PRs, to prevent this PR from becoming too big, and including lots of mechanical changes, making it harder to review.

### CODE REVIEW

File: lib/value/src/kind.rs
Comment (JeanMertz) [inline lib/value/src/kind.rs:33]: Just a note that I started with booleans, but preferred having all fields be `Option`s, since the collection fields needed that anyway. These fields aren't public, but it made for a more consistent internal API.

Also, `size_of::<bool>() == size_of::<Option<()>>()`, so there's no performance loss (not that it matters too much for this type and the way it is used).

File: lib/value/src/kind/accessor.rs
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:36]: I'm not sure if this is correct.

Say we have an object `{thing: int | null, thung: float}` and we do `find_at_path(".(thing | thung)")`. 

I believe this should return `int | float`, since the resulting type at runtime of fetching that path would either be an `int` or a `float`, depending on the value of `thing` at runtime.

This function would currently return `int | null`.

The coalesce needs to fold the types of all the potentially nullable fields together.

File: lib/value/src/kind/accessor.rs
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:72]: Couldn't we add the kind to all fields indicated in the coalesce?

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:72]: We could, but I wouldn't actually call that "correct" either.

I consider coalesced paths to be "start from the first, and progress until one matches", and when inserting, I consider it to be "take the first, no matter what" (which I realize is different from this, so I should update this).

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:36]: Good catch!

File: lib/value/src/kind/comparison.rs
Comment (StephenWakely) [inline lib/value/src/kind/comparison.rs:175]: I think a slightly easier to follow, albeit marginally slower implementation would be:

```rust
let mut count = 0;
if self.is_bytes() {
  count += 1;
}
if self.is_integer(){
  count += 1;
}
if self.is_float(){
  count += 1;
}
if self.is_boolean(){
  count += 1;
}
if self.is_timestamp(){
  count += 1;
}
if self.is_regex() {
  count += 1;
}
if self.is_null(){
  count += 1;
}
if self.is_array(){
  count += 1;
}
if self.is_object(){
  count += 1;
}
count == 1
```

File: lib/value/src/kind/accessor.rs
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:152]: I think this should recurse down all matching fields in the coalesce to update all matching paths.

**Review by StephenWakely — DISMISSED** [review]
A few issues with the coalesce functionality, which is unfortunate since they are it seems hardly ever used, but worth getting right if we decide to continue supporting them!

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:152]: Similar to above, I don't think it should. It should stop at the first matched field, similar to how querying a coalesced path stops at querying the first matched field.

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:175]: Agreed. That is easier to reason about, and performance isn't that much of an issue in this code path.

File: lib/value/src/kind/accessor.rs
Comment (StephenWakely) [reply lib/value/src/kind/accessor.rs:152]: Given this typedef, `{thing: { thang: int } | null, thung: { thang: int } }`.

The typedef for that could represent both objects:

`{ thing: { thang: 32 }, thung: { thang: 43 } }`
`{ thung: { thang: 43 } }`

Calling `del(x, .(thing | thung).thang)` would result in:

`{ thing: {}, thung: { thang: 43 } }`
`{ thung: {} }`

Thereby the typedef for the object should represent the fact that `thang` can be removed from both branches.

If you call `remove_at_path(.(thing | thung).thang)`, currently the result would be: `{ thung: { thang: float }`, implying that `thang` can never be removed from `thung`.

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:152]: Ah yes, you are totally correct. Ugh, these coalesced paths are such a mess to work with. I'm curious if we actually handled this correctly before. Regardless, it should work as expected, so I'll look into this.

Depending on the complexity, I might do it in a follow-up PR so that I'm not blocked on moving the schema work forward for now. It _might_ be the case that I can no longer take a `&mut self` but have to take the `Kind` by value and return a cloned variant, which would be unfortunate, but the price of getting this right. Although it might still be possible to keep the mutable reference, I'll have to see how much the borrow-checker complains (implementing these functions using `&mut self` was quite a bit of work to get right, the old implementation used cloning to work around that).

Comment (StephenWakely) [reply lib/value/src/kind/accessor.rs:152]: I wonder if we could just remove all type defs at the first coalesce and say anything goes after that?

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:152]: That would be a "correct, but non-complete" way to tackle this. I think I'm for this approach to start, we can refine it if we really see a need to do so.

File: lib/value/src/kind/comparison.rs
Comment (pablosichert) [inline lib/value/src/kind/comparison.rs:270]: Where / how do we use the `intersects` method? I'd like to have some context to reason about the collection cases.

**Review by pablosichert — COMMENTED** [review]
Super nice work, @JeanMertz! That was quite some involved logic to unpack. The overall state looks really good to me. There's just a couple of questions I have to clarify my understanding to give a final approval.

File: lib/value/src/kind/comparison.rs
Comment (pablosichert) [reply lib/value/src/kind/comparison.rs:175]: (Might also just use `self.is_bytes() + ... + self.is_object() == 1`.)

File: lib/value/Cargo.toml
Comment (JeanMertz) [reply lib/value/Cargo.toml:4]: Ah yeah, I mostly keep copying this from one VRL lib to another. Will get the correct metadata before I merge this 👍

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:129]: Yeah, that's a nice addition. I briefly contemplated adding doc-tests everywhere, but didn't go that route just yet. I'll see if I can add a few here and there, it's straightforward to do, just takes a bunch of time to get them all documented.

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:269]: I suspect I copied this example from the `remote_at_path` tests, since they are common in structure, just with different results. This one doesn't make sense for `find_at_path` though, since no mutation happens of the original event, will update 👍

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:370]: Yeah, this behaviour is questionable. I'll think on this, and will circle back to this in the follow-up PR that links it up to VRL, if I think there's a better way to handle this (preferably without panicking, and most-likely also without returning an error because there are places in VRL where this method is called that doesn't allow returning an error. Although I might opt to return an error for this method, but then panic at the VRL call-site, if we deem it to be a programming error).

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:478]: It's not a programming error per sé, there are places where people can use negative indexing in paths. But yeah, it is an edge-case, and one we (currently) don't really support.

I can see an update to the type system in the future that would allow us to better track array sizes, and then allowing or rejecting negative indexing more explicitly.

File: lib/value/src/kind/builder.rs
Comment (JeanMertz) [reply lib/value/src/kind/builder.rs:375]: Good suggestion, will update the rest as well.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:41]: Interesting. Not sure, I don't think Clippy warned me of this one, but I might have overlooked the error. I'll take a peek before merging.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:71]: It should, good suggestion.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:87]: I think Clippy doesn't warn here because we take `&mut self`, which often (but not always, such as here) implies the method has side effects, that don't always require the use of the return value.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:149]: It's mostly to be consistent with the existing implementation (i.e. to make all `stdlib` tests pass).

We'll probably want to come up with a better name for the strategy, but I'm not sure what to call it. I'll think about this a bit when I do another pass once the library is integrated into VRL.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:232]: Often, the starting states are mostly there for me to mix-and-match different type states so that there aren't any weird bugs that creep in that happen only for one state, but not the other.

Small chance, and usually this would be better tested with property tests and the like, but it's a small price to pay, for a tiny bit of extra security.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:287]: Good catch. I added these tests before I added that strategy (since it was needed to pass some VRL tests). I updated the test case by copy/pasting the same strategy into all cases, but forgot to add new cases testing the different strategies. Will do so.

File: lib/value/src/kind/collection/exact.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection/exact.rs:28]: I explicitly didn't want to use it because we use(d) it in the old implementation, and while it is cool and all at first, it just comes with an API that is broader than we need, and it caused more confusion than it was worth over the last year+.

The most frequently confusing example I can remember is that bitflags has an `all()` function to initialize the type, but we call it `any()`, as that better fits the mental model of progressive type checking, but every so often, that method popped up in the code because people's editors auto-completed it, and it _works_, but it's confusing to see it being used next to `any()`.

TL;DR: at a surface, it makes sense, but after having used it for more than a year, I found it to be more confusing than worthwhile.

File: lib/value/src/kind/collection/exact.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection/exact.rs:48]: Correct, but as explained above, I explicitly avoided the crate due to my experience with it in the "old" `Kind` implementation.

File: lib/value/src/kind/collection/index.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection/index.rs:43]: I believe in `remove_at_path`, when down-shifting indices.

File: lib/value/src/kind/collection/unknown.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection/unknown.rs:54]: I'm not certain why Clippy doesn't warn in some cases, but will add.

File: lib/value/src/kind/collection/unknown.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection/unknown.rs:83]: It should, I didn't pay much attention to these impls, other than having them for consistency, but it's actually one of the other regrets I have of the old `Kind` impl, and I've since removed these from the follow-up PR. It's just too confusing in everyday use, given the complexities that come from the conceptual model of progressive type checking.

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:175]: `is_*` returns a boolean, not sure how that would work?

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:270]: In VRL, you can search the existing `type_def` in master, or wait for the follow-up PR.

File: lib/value/src/kind/collection/exact.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection/exact.rs:28]: I'll add that we _could_ wrap the `bitflags` type in a newtype, but ultimately, I just don't think the juice is worth the squeeze.

File: lib/value/src/kind/merge.rs
Comment (spencergilbert) [inline lib/value/src/kind/merge.rs:125]: I think elsewhere in the PR you imported this at the top, is there a reason why they weren't all done the same?

File: lib/value/src/kind/insert.rs
Comment (spencergilbert) [inline lib/value/src/kind/insert.rs:31]: Leaf segment == final path segment?

Comment (JeanMertz) [reply lib/value/src/kind/insert.rs:31]: Correct. It's common terminology for tree-like structures, although paths aren't necessarily a tree structure, so if you think it's confusing, let me know and we can change it.

File: lib/value/src/kind/merge.rs
Comment (JeanMertz) [reply lib/value/src/kind/merge.rs:125]: No, will check for consistency.

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:255]: It's not, it's just a random number I picked for testing (arrays can have gaps in their indices, if we don't have any information for a given position (this is what `unknown` details are for in a collection).

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:60]: It is. It's a helper function to allow recursion down below.

File: lib/value/src/kind/collection.rs
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:93]: Yeah, exactly, there's already an `unknown` for initializing. However, I've updated parts of this code in a follow-up PR that's coming today, where I have `from_unknown` to do the initialization, and then there's `unknown` and `unknown_mut` similar to `known` and `known_mut`.

File: lib/value/src/kind/accessor.rs
Comment (spencergilbert) [reply lib/value/src/kind/accessor.rs:60]: Interesting, I don't know if I remember seeing a `fn` declared inside of another `fn` - thanks!

File: lib/value/src/kind/insert.rs
Comment (spencergilbert) [reply lib/value/src/kind/insert.rs:31]: Nope - leaf is fine, just wanted to double check as I mid brain melt at this point :)

File: lib/value/src/kind/comparison.rs
Comment (StephenWakely) [reply lib/value/src/kind/comparison.rs:175]: Perhaps: 

```rust
let a = |b| if b { 1 } else { 0 };

a(self.is_bytes()) + 
a(self.is_integer()) + 
a(self.is_float()) + 
a(self.is_boolean()) + 
a(self.is_timestamp()) + 
a(self.is_regex()) + 
a(self.is_null()) + 
a(self.is_array()) + 
a(self.is_object()) == 1
```

**Comment by fuchsnj** [comment]
I think `"lib/value"` should be added to the workspace members list in the root `Cargo.toml`

File: lib/value/Cargo.toml
Comment (JeanMertz) [reply lib/value/Cargo.toml:4]: Actually, it appears we haven't updated this in other libraries either. I think it's best to delegate this to a separate PR. I created an issue to track this at #11103.

File: lib/value/src/kind/comparison.rs
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs]: After the latest refactor, this code is gone. I haven't added any more doc-tests, but we could do so in the future if we think it adds value.

**Review by pablosichert — APPROVED** [review]
Awesome work, looks good to me.

File: lib/value/src/kind/accessor.rs
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:16]: I'm not convinced these `must_use` annotations are useful since I can't see any scenario where someone would call this function and forget to use the result. 

I think the intention behind this is for situations where the functions success path returns no useful result, but the error path returns an error. A developer could forget to handle the error and think their code is working fine.

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:16]: The documentation (and RFC) has examples that specifically call out any use-case where the result of a function should be used for that function to be useable (e.g. for a pure function that has no side effects).

See: https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute

In fact, there's [ongoing work](https://github.com/rust-lang/rust/issues/89692) to expand this attribute across Rust's stdlib as well (such as `len()` and `is_empty()`).

See also this "[for reviewers](https://std-dev-guide.rust-lang.org/code-considerations/design/must-use.html#for-reviewers)" section, which calls out:

> Look for any legitimate use-cases where `#[must_use]` will cause callers to explicitly ignore values. If these are common then `#[must_use]` probably isn't appropriate.

In this (and other) case(s), calling `find_at_path`, but _not_ using the return value is clearly a programmer bug, and so a valid case to apply `#[must_use]` to.

Personally, I don't mind _too much_ whether we add it or not, but it's a nice tool to have in our toolbox to prevent any bugs in API misuse down the line, and while it does add a bit of clutter, I'm mostly on the side that the clutter is worth the extra safety.

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs]: I decided to reject coalesced nesting for now. We can revisit this later.

File: lib/value/src/kind/accessor.rs
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs]: I've instead opted to reject coalesced path deletion (in this library) for now.

**Pull request merged**
